ArchiSteamFarm/ArchiWebHandler.cs

```diff
@@ -1137,13 +1137,40 @@ internal sealed class ArchiWebHandler : IDisposable {
 				return null;
 			}
 
-			const string request = "/account/validatewalletcode";
+			const string requestValidateCode = "/account/validatewalletcode";
 
 			// Extra entry for sessionID
 			Dictionary<string, string> data = new Dictionary<string, string>(2) { { "wallet_code", key } };
 
-			Steam.RedeemWalletResponse response = await UrlPostToJsonObjectWithSession<Steam.RedeemWalletResponse>(SteamStoreURL, request, data).ConfigureAwait(false);
-			return response == null ? ((EResult Result, EPurchaseResultDetail? PurchaseResult)?) null : (response.Result, response.PurchaseResultDetail);
+			Steam.RedeemWalletResponse responseValidateCode = await UrlPostToJsonObjectWithSession<Steam.RedeemWalletResponse>(SteamStoreURL, requestValidateCode, data).ConfigureAwait(false);
+			if (responseValidateCode == null) {
+				return null;
+			}
+
+			if (responseValidateCode.Result != EResult.OK) {
+				return (responseValidateCode.Result, responseValidateCode.PurchaseResultDetail);
+			}
+
+			if (responseValidateCode.KeyDetails == null) {
+				Bot.ArchiLogger.LogNullError(nameof(responseValidateCode.KeyDetails));
+				return null;
+			}
+
+			if (responseValidateCode.WalletCurrencyCode != responseValidateCode.KeyDetails.CurrencyCode) {
+				const string requestCheckFunds = "/account/createwalletandcheckfunds";
+				Steam.EResultResponse responseCheckFunds = await UrlPostToJsonObjectWithSession<Steam.EResultResponse>(SteamStoreURL, requestCheckFunds, data).ConfigureAwait(false);
+				if (responseCheckFunds == null) {
+					return null;
+				}
+
+				if (responseCheckFunds.Result != EResult.OK) {
+					return (responseCheckFunds.Result, null);
+				}
+			}
+
+			const string requestConfirmRedeem = "/account/confirmredeemwalletcode";
+			Steam.RedeemWalletResponse responseConfirmRedeem = await UrlPostToJsonObjectWithSession<Steam.RedeemWalletResponse>(SteamStoreURL, requestConfirmRedeem, data).ConfigureAwait(false);
+			return responseConfirmRedeem == null ? ((EResult Result, EPurchaseResultDetail? PurchaseResult)?) null : (responseConfirmRedeem.Result, EPurchaseResultDetail.NoDetail);
```

Comment:
When using `?`, prefer to have expected case first and second one after, so change order here to have `!= null`.

Reply:
You still shouldn't assume `EPurchaseResultDetail` here. Return the one that the function returned, you have it available now.